// slydpay_mantis/index.js
const rp = require('request-promise');
let uuid = require('uuid4');
let moment = require('moment');

module.exports = {
    slydpay_invoice: async function (data, create_invoice_callback) {

        let payment_method = data.payment_method;
        let payment_options = ["VODAFONE_CASH", "AIRTEL_MONEY", "MTN_MONEY", "VODAFONE_CASH_PROMPT"];

        for (let i = 0; i < payment_options.length; i++) {
            let supp = payment_options[i];

            console.log(supp === payment_method);

            if (payment_method === supp) {
                let invoice = get_create_and_send_invoice(data.values, data, supp, data.payment_type);
                toReturn = invoice;
            }
        }

        let resp = await toReturn.then(resp => {
            return resp;
        }).catch(err => {
            console.log(err);
            return err;
        });

        create_invoice_callback(null, resp);
    }
};

const get_create_and_send_invoice = async (values, data, payment_method,payment_type) => {

    let body = {
        "emailOrMobileNumber": values.emailOrMobileNumber,
        "merchantKey": values.merchantKey,
        "amount": data.amount,
        "description": data.payment_desc,
        "orderCode": uuid(),
        "sendInvoice": true,
        "payOption": payment_method,
        "customerName": data.customerName,
        "customerMobileNumber": data.phone_number
    };

    let options = {
        url: 'https://app.slydepay.com.gh/api/merchant/invoice/create',
        method: "POST",
        headers: {
            "content-type": "application/json",
        },
        json: body
    };

    let response = rp(options);

    await response.then(d => {

        (function payment_status() {
            const statusInfo = check_payment_status(d, values);
            response = statusInfo;

            statusInfo.then(status_data => {
                switch (status_data.result) {
                    case "NEW":
                        console.log("new")
                        setTimeout(payment_status, 10000);
                        break
                    case "PENDING":
                        console.log("pending")
                        confirm_transaction(d, values)
                        setTimeout(payment_status, 10000);
                        break
                    case "CONFIRMED":
                        (function confirmTransaction() {
                            const transaction_info = confirm_transaction(d, values);
                            transaction_info.then(inf => {

                                if (inf.result !== "CONFIRMED") {
                                    confirm_transaction(d, values);
                                    setTimeout(confirmTransaction, 10000);
                                }

                                send_payment_to_callback(inf, data, payment_method, payment_type);
                            }).catch(e => {
                                console.log(e.message);
                            })
                        })()
                        break
                    case "DISPUTED":
                        console.log("DISPUTED")
                        send_payment_to_callback(status_data, data, payment_method, payment_type);
                        break
                    case "CANCELLED":
                        console.log("CANCELLED")
                        send_payment_to_callback(status_data, data, payment_method, payment_type);
                        break
                    default:
                        console.log("nothing");
                }
            }).catch(err => {
                console.log(err.message);
            })
        })()
    });

    return response;
};

const send_payment_to_callback = (payment_state_info, data, payment_method, payment_type) => {
    let body = {
        phone_number: data.phone_number,
        amount: data.amount,
        date: new Date(moment().date()),
        time: new Date(moment().time),
        provider: payment_method,
        reference: `${payment_type}${create_ref_id(10)}`,
        status: payment_state_info.result,
        product_id: data.product_id
    };

    console.log("data to callback: ", body);

    let options = {
        url: data.callback_url,
        method: "POST",
        headers: {
            "content-type": "application/json",
        },
        json: body
    }

    let response = rp(options);
    response.then(to_callback => {
        console.log(to_callback);
    });

    return response;
}

function create_ref_id(length) {
    let result           = '';
    let characters       = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
    let charactersLength = characters.length;
    for ( let i = 0; i < length; i++ ) {
        result += characters.charAt(Math.floor(Math.random() * charactersLength));
    }
    return result;
}

const check_payment_status = (resp_data, req_data) => {

    console.log(resp_data);
    //console.log(body);

    let body = {
        "emailOrMobileNumber": req_data.emailOrMobileNumber,
        "merchantKey": req_data.merchantKey,
        "orderCode": resp_data.result.orderCode,
        "payToken": resp_data.result.payToken,
        "confirmTransaction": true
    };

    let options = {
        url: 'https://app.slydepay.com.gh/api/merchant/invoice/checkstatus',
        method: "POST",
        headers: {
            "content-type": "application/json",
        },
        json: body
    }

    const response = rp(options);

    return response
};

const confirm_transaction = (resp_data, req_data) => {
    let body = {
        "emailOrMobileNumber": req_data.emailOrMobileNumber,
        "merchantKey": req_data.merchantKey,
        "orderCode": resp_data.result.orderCode,
        "payToken": resp_data.result.payToken,
    };

    let options = {
        url: 'https://app.slydepay.com.gh/api/merchant/transaction/confirm',
        method: "POST",
        headers: {
            "content-type": "application/json",
        },
        json: body
    }

    const response = rp(options);

    return response;
};
